ykchalresp.1
.\" Copyright (c) 2011-2014 Yubico AB
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\"
.\"     * Redistributions of source code must retain the above copyright
.\"       notice, this list of conditions and the following disclaimer.
.\"
.\"     * Redistributions in binary form must reproduce the above
.\"       copyright notice, this list of conditions and the following
.\"       disclaimer in the documentation and/or other materials provided
.\"       with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" The following commands are required for all man pages.
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.TH ykchalresp "1" "Febuary 2011" "yubikey-personalization"
.SH NAME
ykchalresp - Perform challenge-response operation with YubiKey
.SH SYNOPSIS
.B ykchalresp
[\fI-1\fR | \fI-2\fR] [\fI-H\fR] [\fI-Y\fR] [\fI-N\fR] [\fI-x\fR] [\fI-v\fR] [\fI-6\fR] [\fI-8\fR] [\fI-t\fR] [\fI-V\fR] [\fI-h\fR]
.SH DESCRIPTION
.PP
Send a challenge to a YubiKey, and read the response.  The YubiKey can be configured
with two different C/R modes -- the standard one is a 160 bits HMAC-SHA1, and the other
is a YubiKey OTP mimicing mode, meaning two subsequent calls with the same challenge
will result in different responses.
.SH OPTIONS
.TP
\fB\-1\fR
send the challenge to slot 1.  This is the default.
.TP
\fB\-2\fR
send the challenge to slot 2.
.TP
\fB\-H\fR
send a 64 byte HMAC challenge.  This is the default.
.TP
\fB\-Y\fR
send a 6 byte Yubico OTP challenge.
.TP
\fB\-N\fR
non-blocking mode -- abort if the YubiKey is configured to require a key press before
sending the response.
.TP
\fB\-x\fR
challenge is hex encoded.
.TP
\fB\-v\fR
enable verbose mode.
.TP
\fB\-6\fR
output the response in OATH format, 6 digits.
.TP
\fB\-8\fR
output the response in OATH format, 8 digits.
.TP
\fB\-t\fR
use current time as challenge instead of reading challenge from command line (as in default TOTP mode, seconds since 1970-01-01 00:00:00 / 30 encoded as an 8 byte challenge).
.TP
\fB\-V\fR
print tool version and exit.

.SH EXAMPLE
The YubiKey challenge-response operation can be demonstrated using the
\fBNIST PUB 198 A.2\fR test vector.
.P
First, program a YubiKey with the test vector :
.HP
.nf
$ \fBykpersonalize \-2 \-ochal\-resp \-ochal\-hmac \-ohmac\-lt64 \-a 303132333435363738393a3b3c3d3e3f40414243\fR
 ...
Commit? (y/n) [n]: y
$
.fi
.HP
Now, send the NIST test challenge to the YubiKey and verify the result matches the
expected :
.HP
.nf
$ \fBykchalresp \-2 'Sample #2'\fR
0922d3405faa3d194f82a45830737d5cc6c75d24
$
.fi

.SH BUGS
Report ykchalresp bugs in
.URL "https://github.com/Yubico/yubikey-personalization/issues" "the issue tracker"
.SH "SEE ALSO"
The
.URL "https://developers.yubico.com/yubikey-personalization/" "ykpersonalize home page"
.PP
YubiKeys can be obtained from
.URL "http://www.yubico.com/" "Yubico" "."


NEWS
Yubikey-personalize NEWS -- History of user-visible changes.     -*- outline -*-

* Version 1.15.3 (released 2014-09-04)

** Fix URLs for opensource.y.com -> developers.y.com move.

** Whitelist firmware version 3.3 and detect new PIDs.

* Version 1.15.2 (released 2014-07-30)

** Whitelist firmware version 2.5

** Read key when importing configuration.

** Fix formatting error in information about what is written to key.

** Check return codes when doinf NDEF writes.

* Version 1.15.1 (released 2014-03-12)

** Add -6 and -8 flags to ykchalresp to output the response as a truncated
OATH code. Thanks to joernchen of Phenoelit <joernchen@phenoelit.de>

** Add -t to ykchalresp to use current time / 30 as challenge in TOTP
fashion. Thanks to joernchen of Phenoelit <joernchen@phenoelit.de>

** Fix for a crash in the libusb-0.1 backend.
Thanks to Antti Keränen <detegr@gmail.com>

* Version 1.15.0 (released 2014-01-10)

** add new Neo version 3.2 as supported.

** let ykp_AES_key_from_password() return error if it can't get a random
salt.

** let -a without option mean ask for raw key on STDIN.

** let no -a given mean generate a random key

** add ykp_AES_key_from_raw() and ykp_HMAC_key_from_raw() that sets the
key from a byte array.

** remove -osalt= and possibility to generate key from password.

** add YK_ENODATA that's returned when no data is returned from
the device and no other underlying error is set.

* Version 1.14.2 (released 2013-12-30)

** let -m take challenge-response timeout and autoeject timeout
supplied with : beteen, so mode:cr-timeout:autoeject-timeout

** make sure to always wait 1 millisecond between writes to the key
newer platforms triggered a race where writes to the key could
happen in a very short (about 0.4 milliseconds) timespan and would
not be processed.

** minor fixes for the printing format.

* Version 1.14.1 (released 2013-09-16)

** Fix breakage with latest automake.

* Version 1.14.0 (released 2013-07-04)

** Fixups of import/export.
Add targetConfig to show in which slot a configuration is intended.
Possible memory leaks on error conditions.

** Add -d switch to ykpersonalize for dry-run.

** Add ykp_clear_config() for clearing configuration flags.

** Add getter functions for all configuration flags.

** Add -V to all tools to output version.

** Add ykp_get_acccode_type() and ykp_set_acccode_type()
Only to do with export, showing where the access code came from
in the ycfg.

** Add -1 and -2 options to ykinfo to show programming state.

* Version 1.13.0 (released 2013-04-22)

** Added import/export of ycfg-json format.
Invoke with -fjson to -s or -i
Add exported functions ykp_export_config() and ykp_import_config()

** Fixup output of flags when using ykp_write_config()

** Add binary builds for mac.

** Minor cleanups noticed during debian packaging.

* Version 1.12.0 (released 2013-03-14)

** Recognize firmwares 2.4 and 3.1.

** Add support for setting the new extflag LED_INV
When set the behaviour of the led on the YubiKey is inversed.

* Version 1.11.3 (released 2013-01-09)

** Fixup of broken release.

* Version 1.11.2 (released 2013-01-09)

** Fix a bug where writing a NDEF with unknown prefix ended up writing invalid
data to the YubiKey NEO. Wrote prefix as 0x24 instead of 0x00.

** Don't allow opening a YubiKey if there's more than one present in the system.

** Fix shared linking of ykinfo and ykchalresp.

* Version 1.11.1 (released 2012-12-21)

** Implement ykusb_strerror() on windows.

** Fix a bug where a YubiKey would fail to be recognized if there was
another device from Yubico (vendor id 1050) inserted and looked at
before in the device chain.

** Fix a bug where you could only set 8 bytes of the public id with
the command line tool, now all 16 bytes can be set.

** Documentation updates and fixes.

* Version 1.11.0 (released 2012-12-12)

** Added version symbols and functions.
The header file is "ykpers-version.h" and it contains the following
symbols and functions: YKPERS_VERSION_STRING, YKPERS_VERSION_NUMBER,
YKPERS_VERSION_MAJOR, YKPERS_VERSION_MINOR, YKPERS_VERSION_PATCH,
ykpers_check_version.  Requested by Giampaolo Bellini
<iw2lsi@gmail.com>.

* Version 1.10.0 (released 2012-12-11)

** Support for the new productId of the production Neo.
Has ProducId 0x110, 0x111 or 0x112 depending on mode (see the notes about
-m and device_config).

** Add support for SLOT_NDEF2.
Use SLOT_NDEF to emit slot 1 as NDEF or SLOT_NDEF2 to emit slot 2.
This also adds the function yk_write_ndef2() that takes a slot parameter.

** Add -m flag for ykpersonalize, set usb mode of YubiKey NEO.
0 means pure YubiKey mode, 1 means pure CCID mode and 2 means YubiKey/CCID
composite mode. Add 80 to set EJECT_FLAG.
To use this with the api, see the functions:
ykp_alloc_device_config(), ykp_free_device_config(), ykp_set_device_mode(),
ykp_set_device_chalresp_timeout(), ykp_set_device_autoeject_time() and
yk_write_device_config().

** Add -S flag for ykpersonalize, set the scanmap of the YubiKey NEO.
Take an 90 character string describing 45 scancodes. See man page for more
info. To use this with the api see yk_write_scan_map().

** In the api add ykp_ndef_as_text() to export the text from a YK_NDEF structure.

** Higher timeout for configuration writes as in particular swap can take
longer than 600 ms.

* Version 1.9.0 (released 2012-11-08)

** ykinfo: New tool to print information about YubiKey.

** ykpersonalize: Add -z flag to zap configuration on YubiKey.

** Fix PBKDF2 implementation.

* Version 1.8.2 (released 2012-10-17)

** Add udev rules files to packed distribution.

* Version 1.8.1 (released 2012-10-17)

** Memory leak fixes and potential crash fixes in osx backend.

** Error reporting fixes in osx backend, reporting correct errors and
better errors.

** Provide new another udev permissions file that works on udev version
greater than 188. Autodetects from configure which to use. Thanks to
Maxim Burgerhout <maxim@wzzrd.com>.


ykchalresp.c
/* -*- mode:C; c-file-style: "bsd" -*- */
/*
 * Copyright (c) 2011-2013 Yubico AB.
 * All rights reserved.
 *
 * Author : Fredrik Thulin <fredrik@yubico.com>
 *
 * Some basic code copied from ykpersonalize.c.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above
 *       copyright notice, this list of conditions and the following
 *       disclaimer in the documentation and/or other materials provided
 *       with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>
#include <unistd.h>
#include <time.h>

#include <yubikey.h>
#include <ykdef.h>
#include <ykcore.h>
#include <ykstatus.h>
#include <ykpers-version.h>

const char *usage =
	"Usage: ykchalresp [options] [challenge]\n"
	"\n"
	"Options :\n"
	"\n"
	"\t-1        Send challenge to slot 1. This is the default.\n"
	"\t-2        Send challenge to slot 2.\n"
	"\t-H        Send a 64 byte HMAC challenge. This is the default.\n"
	"\t-Y        Send a 6 byte Yubico challenge.\n"
	"\t-N        Abort if Yubikey requires button press.\n"
	"\t-x        Challenge is hex encoded.\n"
	"\t-t        Time based challenge (for TOTP)\n"
	"\t-6        Output 6 digit HOTP/TOTP code\n"
	"\t-8        Output 8 digit HOTP/TOTP code\n"
	"\n"
	"\t-v        verbose\n"
	"\t-V        tool version\n"
	"\t-h        help (this text)\n"
	"\n"
	"\n"
	;
const char *optstring = "1268xvhHtYNV";

static void report_yk_error(void)
{
	if (yk_errno) {
		if (yk_errno == YK_EUSBERR) {
			fprintf(stderr, "USB error: %s\n",
				yk_usb_strerror());
		} else {
			fprintf(stderr, "Yubikey core error: %s\n",
				yk_strerror(yk_errno));
		}
	}
}

static int parse_args(int argc, char **argv,
	       int *slot, bool *verbose,
	       unsigned char **challenge, unsigned int *challenge_len,
	       bool *hmac, bool *may_block, bool *totp, int *digits,
	       int *exit_code)
{
	int c;
	bool hex_encoded = false;

	while((c = getopt(argc, argv, optstring)) != -1) {
		switch (c) {
		case '1':
			*slot = 1;
			break;
		case '2':
			*slot = 2;
			break;
		case 'H':
			*hmac = true;
			break;
		case 'N':
			*may_block = false;
			break;
		case 't':
			*totp = true;
			*hmac = true;
			break;
		case '6':
			*digits = 6;
			break;
		case '8':
			*digits = 8;
			break;
		case 'Y':
			*hmac = false;
			*totp = false;
			break;
		case 'x':
			hex_encoded = true;
			break;
		case 'v':
			*verbose = true;
			break;
		case 'V':
			fputs(YKPERS_VERSION_STRING "\n", stderr);
			*exit_code = 0;
			return 0;
		case 'h':
		default:
			fputs(usage, stderr);
			*exit_code = 0;
			return 0;
		}
	}

	if ((optind >= argc && !*totp) || (optind < argc && *totp)) {
		/* No challenge */
		fputs(usage, stderr);
		return 0;
	}
	if (*totp && *hmac) {
		unsigned int t_counter;
		static unsigned char t_buf[8];
		t_counter = (int) time(NULL);
		t_counter = t_counter / 30;
		memset(t_buf, 0, sizeof(t_buf));
		t_buf[7] = t_counter & 0x000000ff;
		t_buf[6] = (t_counter & 0x0000ff00) >> 8;
		t_buf[5] = (t_counter & 0x00ff0000) >>16;
		t_buf[4] = (t_counter & 0xff000000) >>24;
		*challenge = (unsigned char *) &t_buf;
		*challenge_len = 8;
	}
	else if (hex_encoded) {
		static unsigned char decoded[SHA1_MAX_BLOCK_SIZE];

		size_t strl = strlen(argv[optind]);

		if (strl > sizeof(decoded) * 2) {
			fprintf(stderr, "Hex-encoded challenge too long (max %lu chars)\n",
				sizeof(decoded) * 2);
			return 0;
		}

		if (strl % 2 != 0) {
			fprintf(stderr, "Odd number of characters in hex-encoded challenge\n");
			return 0;
		}

		memset(decoded, 0, sizeof(decoded));

		if (yubikey_hex_p(argv[optind])) {
			yubikey_hex_decode((char *)decoded, argv[optind], sizeof(decoded));
		} else {
			fprintf(stderr, "Bad hex-encoded string '%s'\n", argv[optind]);
			return 0;
		}
		*challenge = (unsigned char *) &decoded;
		*challenge_len = strl / 2;
	} else {
		*challenge = (unsigned char *) argv[optind];
		*challenge_len = strlen(argv[optind]);
	}

	return 1;
}

static int check_firmware(YK_KEY *yk, bool verbose)
{
	YK_STATUS *st = ykds_alloc();

	if (!yk_get_status(yk, st)) {
		ykds_free(st);
		return 0;
	}

	if (verbose) {
		printf("Firmware version %d.%d.%d\n",
		       ykds_version_major(st),
		       ykds_version_minor(st),
		       ykds_version_build(st));
		fflush(stdout);
	}

	if (ykds_version_major(st) < 2 ||
	    (ykds_version_major(st) == 2
	     && ykds_version_minor(st) < 2)) {
		fprintf(stderr, "Challenge-response not supported before YubiKey 2.2.\n");
		ykds_free(st);
		return 0;
	}

	free(st);
	return 1;
}

static int challenge_response(YK_KEY *yk, int slot,
		       unsigned char *challenge, unsigned int len,
		       bool hmac, bool may_block, bool verbose, bool totp, int digits )
{
	unsigned char response[SHA1_MAX_BLOCK_SIZE];
	unsigned char output_buf[(SHA1_MAX_BLOCK_SIZE * 2) + 1];
	int yk_cmd;
	unsigned int expect_bytes = 0;
	unsigned int offset;
	unsigned int bin_code;
	memset(response, 0, sizeof(response));
	memset(output_buf, 0, sizeof(output_buf));

	if (verbose) {
		fprintf(stderr, "Sending %i bytes %s challenge to slot %i\n", len, (hmac == true)?"HMAC":"Yubico", slot);
	}

	switch(slot) {
	case 1:
		yk_cmd = (hmac == true) ? SLOT_CHAL_HMAC1 : SLOT_CHAL_OTP1;
		break;
	case 2:
		yk_cmd = (hmac == true) ? SLOT_CHAL_HMAC2 : SLOT_CHAL_OTP2;
		break;
	default:
		return 0;
	}

	if(! yk_challenge_response(yk, yk_cmd, may_block, len,
				challenge, sizeof(response), response)) {
		return 0;
	}

	/* HMAC responses are 160 bits, Yubico 128 */
	expect_bytes = (hmac == true) ? 20 : 16;

	if(digits && hmac){
		offset   =  response[19] & 0xf ;
		bin_code = (response[offset]  & 0x7f) << 24
			| (response[offset+1] & 0xff) << 16
			| (response[offset+2] & 0xff) <<  8
			| (response[offset+3] & 0xff) ;
		if(digits == 8){
			bin_code = bin_code % 100000000;
			printf("%08u\n", bin_code);
			return 1;
		}
		bin_code = bin_code % 1000000;
		printf("%06i\n", bin_code);
		return 1;
	}
	if (hmac) {
		yubikey_hex_encode((char *)output_buf, (char *)response, expect_bytes);
	} else {
		yubikey_modhex_encode((char *)output_buf, (char *)response, expect_bytes);
	}
	printf("%s\n", output_buf);

	return 1;
}

int main(int argc, char **argv)
{
	YK_KEY *yk = 0;
	bool error = true;
	int exit_code = 0;

	/* Options */
	bool verbose = false;
	bool hmac = true;
	bool may_block = true;
	bool totp = false;
	int digits = 0;
	unsigned char *challenge;
	unsigned int challenge_len;
	int slot = 1;

	yk_errno = 0;

	if (! parse_args(argc, argv,
			 &slot, &verbose,
			 &challenge, &challenge_len,
			 &hmac, &may_block, &totp, &digits,
			 &exit_code))
		exit(exit_code);

	if (!yk_init()) {
		exit_code = 1;
		goto err;
	}

	if (!(yk = yk_open_first_key())) {
		exit_code = 1;
		goto err;
	}

	if (! check_firmware(yk, verbose)) {
		exit_code = 1;
		goto err;
	}

	if (! challenge_response(yk, slot,
				 challenge, challenge_len,
				 hmac, may_block, verbose, totp, digits)) {
		exit_code = 1;
		goto err;
	}

	exit_code = 0;
	error = false;

err:
	if (error || exit_code != 0) {
		report_yk_error();
	}

	if (yk && !yk_close_key(yk)) {
		report_yk_error();
		exit_code = 2;
	}

	if (!yk_release()) {
		report_yk_error();
		exit_code = 2;
	}

	exit(exit_code);
}
